### DISCUSSION TITLE
Unexpected Logout

### DISCUSSION THREAD

**Discussion by Sepehrworklife** [aminalaee/sqladmin#443] [Q&A] [answered]
Hello! I implemented AuthenticateBackEnd as the documentation said.
The problem is when I walk around pages suddenly page will redirect to login.
I checked the responses, at first it will return 302 Found and then redirected to the login page with a 200 status code. 
My authenticate function is just like a document:
`return "token" in request. session` 
It should be mentioned that everything is okay in the first minutes then the problem occurs. There is no specific time too.

If it helps:
- Proxy server is Apache 
- SSL enabled. (works fine)
- Didn't add session middleware manually myself

I would be grateful if you could help me.

**Comment by aminalaee** [comment, marked as answer]
Hey,
I doubt it has anything to do with time, the SessionMiddleware used by default should be valid for two weeks.

Can you post the flow example of the request?
You are trying to access something like /admin/users and you are redirected to login page?
I think the 302 is when `token" in request. session` is False, so you are redirected to the login, but it's not clear why it's happening.

**Reply by Sepehrworklife** [reply]
Yes, I guess that too.
The "token" returns False somehow and I can't recognize why?

```
class AdminAuthenticate(AuthenticationBackend):
    async def login(self, request: Request) -> bool:
        form = await request.form()
        username, password = form.get("username"), form.get("password")
        if AuthModel().authenticate_user(username=username, password=password):
            user = UsersModel().get_user(username)
            user = {
                "username": user.username,
                "password": user.password,
                "is_superuser": user.is_superuser,
            }
            if user["is_superuser"]:
                # I tried {"token": True} before but not work either
                token = AuthModel().create_access_token(user)
                request.session.update({"token": token})
                return True
        return False

    async def logout(self, request: Request) -> bool:
        request.session.clear()
        return True

    async def authenticate(self, request: Request) -> bool:
        return "token" in request.session
```


Requests:
```
https://backdrop.example.com/moderator/loginGET 200 3/10/2023 12:41:41 PM
N/A
https://backdrop.example.com/moderator/login POST 302 3/10/2023 12:41:47 PM
N/A
https://backdrop.example.com/moderator/ GET 200 3/10/2023 12:41:48 PM
N/A


DELETED A LOT OF REQUESTS SO THE PROBLEM OCCURS.


https://backdrop.example.com/moderator/products/list GET 200 3/10/2023 12:42:05 PM
N/A

https://example.com/_next/image?url=%2Fuploads%2Fimages%2Flogo.png&w=64&q=75GET3043/10/2023 12:42:06 PM
N/A
https://backdrop.example.com/moderator/messages/listGET2003/10/2023 12:42:07 PM
N/A

https://example.com/_next/image?url=%2Fuploads%2Fimages%2Flogo.png&w=64&q=75GET3043/10/2023 12:42:07 PM
N/A
https://backdrop.example.com/moderator/news/listGET2003/10/2023 12:42:08 PM
N/A

https://example.com/_next/image?url=%2Fuploads%2Fimages%2Flogo.png&w=64&q=75GET3043/10/2023 12:42:09 PM
N/A
https://backdrop.example.com/moderator/files/list/GET2003/10/2023 12:42:10 PM
N/A

https://example.com/_next/image?url=%2Fuploads%2Fimages%2Flogo.png&w=64&q=75GET3043/10/2023 12:42:10 PM
N/A
https://backdrop.example.com/moderator/destinations/listGET2003/10/2023 12:42:11 PM
N/A

https://example.com/_next/image?url=%2Fuploads%2Fimages%2Flogo.png&w=64&q=75GET3043/10/2023 12:42:12 PM
N/A
https://backdrop.example.com/moderator/files/list/GET2003/10/2023 12:42:18 PM
N/A

https://example.com/_next/image?url=%2Fuploads%2Fimages%2Flogo.png&w=64&q=75GET3043/10/2023 12:42:19 PM
N/A
https://backdrop.example.com/moderator/users/list GET 302 3/10/2023 12:42:20 PM
N/A
https://backdrop.example.com/moderator/login GET 200 3/10/2023 12:42:20 PM
N/A


```

**Reply by aminalaee** [reply]
have you tried it without the Apache reverse proxy? I think that can at least narrow down the issue.

**Reply by Sepehrworklife** [reply]
Thank you so much for your kind replies.
I tried it without the Apache Reverse Proxy ( IP: PORT without SSL ) also the same behavior happened.
I taught somehow it is related to the SQLAlchemy warnings, although after fixing them nothing changed.

**Reply by aminalaee** [reply]
> Thank you so much for your kind replies.

No worries :)

Can you also post how you are using the `AdminAuthenticate` in Admin?
You don't generate random secret key by any  chance?

**Reply by Sepehrworklife** [reply]
Actually, I am:))
```
auth_backend = AdminAuthenticate(secret_key=secrets.token_hex(32))
```
Is it wrong?
I read about it in an article that this way is better than generating statically.

**Reply by aminalaee** [reply]
Ah that's the issue.
Each time your app is restarted, or I think when a worker is killed and restarted automatically in Gunicorn, it will generate a random secret here and previous sessions will be invalid.
I think this will cause a lot of issues, depending on how you deploy your app.

**Reply by Sepehrworklife** [reply]
Ahh, I see.
 I was struggling with this issue for several weeks.
Thank you so much for your helps 🙏

**Comment by Sepehrworklife** [comment]
With @aminalaee helps I recognized that this issue occurred because the session secret key generated every time a gunicorn worker killed or restarted.
When I used a static secret key the problem has been solved.

Thanks to @aminalaee
